rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1559 @@
+    /// if there are any underlying errors from the AIO RPC protocol.
+    pub async fn update_device_plus_endpoint_status(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        status: DeviceStatus,
+        timeout: Duration,
+    ) -> Result<Device, Error> {
+        let status_payload = adr_name_gen::UpdateDeviceStatusRequestPayload {
+            device_status_update: status.into(),
+        };
+        let update_device_status_request =
+            adr_name_gen::UpdateDeviceStatusRequestBuilder::default()
+                .payload(status_payload)
+                .map_err(ErrorKind::from)?
+                .topic_tokens(Self::get_topic_tokens(device_name, inbound_endpoint_name))
+                .timeout(timeout)
+                .build()
+                .map_err(ErrorKind::from)?;
+        let response = self
+            .update_device_status_command_invoker
+            .invoke(update_device_status_request)
+            .await
+            .map_err(ErrorKind::from)?;
+        Ok(response.payload.updated_device.into())
+    }
+
+    /// Starts observation of a Device's updates from the Azure Device Registry service.
+    ///
+    /// Note: On cleanup, unobserve should always be called so that the service knows to stop sending notifications.
+    ///
+    /// # Arguments
+    /// * `device_name` - The name of the Device.
+    /// * `inbound_endpoint_name` - The name of the inbound endpoint.
+    /// * `timeout` - The duration until the Client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns the [`DeviceUpdateObservation`] if the observation was started successfully or [`Error`].
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`DuplicateObserve`](ErrorKind::DuplicateObserve)
+    /// if the device is already being observed.
+    ///
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if timeout is 0.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if there are any underlying errors from the AIO RPC protocol.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if device, inbound endpoint are invalid.
+    ///
+    /// [`struct@Error`] of kind [`ObservationError`](ErrorKind::ObservationError)
+    /// if the observation was not accepted by the service.
+    pub async fn observe_device_update_notifications(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        timeout: Duration,
+    ) -> Result<DeviceUpdateObservation, Error> {
+        let receiver_id = Self::hash_device_endpoint(&device_name, &inbound_endpoint_name);
+        let rx = self
+            .device_update_notification_dispatcher
+            .register_receiver(receiver_id.clone())
+            .map_err(ErrorKind::from)?;
```

Comment:
Design note, and not one we need to use here by any means, but you can do the error kind conversion nicely if there's a wrapper function. So, an internal function that returns `Result<X, ErrorKind>` can automatically convert using the `?` operator, and then the outer function that returns `Result<X, Error>` can also automatically convert using the `?` operator.

But mapping the error everywhere works too. I don't think this alone is enough reason to have separate internal functions, but if you have other reasons as well, it becomes very convenient.